motion/thruster_allocator/src/allocator_ros.cpp

```diff
@@ -0,0 +1,110 @@
+#include <vortex_msgs/msg/thruster_forces.hpp>
+
+#include <chrono>
+#include <functional>
+
+using namespace std::chrono_literals;
+
+ThrusterAllocator::ThrusterAllocator()
+    : Node("thruster_allocator_node"),
+      pseudoinverse_allocator_(Eigen::MatrixXd::Zero(3, 4)) {
+  subscription_ = this->create_subscription<geometry_msgs::msg::Wrench>(
+      "thrust/wrench_input", 1,
+      std::bind(&ThrusterAllocator::wrench_callback, this,
+                std::placeholders::_1));
+
+  publisher_ = this->create_publisher<vortex_msgs::msg::ThrusterForces>(
+      "thrust/thruster_forces", 1);
+
+  timer_ = this->create_wall_timer(
+      100ms, std::bind(&ThrusterAllocator::timer_callback, this));
+
+  Eigen::MatrixXd thrust_configuration_pseudoinverse;
+  calculateRightPseudoinverse(thrust_configuration,
+                              thrust_configuration_pseudoinverse);
+
+  // Set T_pinv in pseudoinverse_allocator_
+  pseudoinverse_allocator_.T_pinv = thrust_configuration_pseudoinverse;
+
+  body_frame_forces_.setZero();
+}
+
+/**
+ * @brief This function is called by the timer and calculates the allocated
+ * thrust based on the body frame forces. It then saturates the output vector
+ * between min and max values and publishes the thruster forces to the topic
+ * "thrust/thruster_forces". If the calculated thruster forces vector is
+ * invalid, it logs an error message. If the thruster forces vector required
+ * saturation, it logs a warning message.
+ */
+
+void ThrusterAllocator::timer_callback() {
+  Eigen::VectorXd thruster_forces =
+      pseudoinverse_allocator_.calculateAllocatedThrust(body_frame_forces_);
+
+  if (isInvalidMatrix(thruster_forces)) {
+    RCLCPP_ERROR(get_logger(), "ThrusterForces vector invalid");
+    return;
+  }
+
+  // Saturates output vector between min and max values
+  if (!saturateVectorValues(thruster_forces, -100, 100)) {
+    RCLCPP_WARN(get_logger(), "Thruster forces vector required saturation.");
+  }
+
+  vortex_msgs::msg::ThrusterForces msg_out;
+  arrayEigenToMsg(thruster_forces, msg_out);
+  // Number of thrusters = 4
+  for (int i = 0; i < 4; i++) {
+    msg_out.thrust[i] *= direction_[i];
+  }
+  publisher_->publish(msg_out);
+}
+
+/**
+ * @brief Callback function for the wrench input subscription. It extracts the
+ * surge, sway and yaw values from the received geometry_msgs::msg::Wrench
+ * message and stores them in the body_frame_forces_ Eigen vector. If the wrench
+ * vector is invalid (contains NaN or Inf values), it logs an error message and
+ * returns.
+ * @param msg The received geometry_msgs::msg::Wrench message.
+ */
+void ThrusterAllocator::wrench_callback(const geometry_msgs::msg::Wrench &msg) {
+  body_frame_forces_(0) = msg.force.x;  // surge
+  body_frame_forces_(1) = msg.force.y;  // sway
+  body_frame_forces_(2) = msg.torque.z; // yaw
+  if (!healthyWrench(body_frame_forces_)) {
+    RCLCPP_ERROR(get_logger(), "ASV wrench vector invalid, ignoring.");
+    return;
+  }
+}
```

Comment:
Logic flow here: 

You set the member variable body_frame_forces_ first, then check if it is invalid. Since the check happens after the internal state update, you don't really prevent a bad wrench msg to be set? (You print "ignoring" but it won't be ignored as far as I can tell)

Reply:
I guess the two alternatives would be to either
1. Set body_frame_forces_ to zero
2. Don't update body_frame_forces_ at all

Reply:
Putting the if-statement first, before setting body_frame_forces_ should fix this right?

Reply:
Yes, if you check msg directly and not body_frame_forces_.